### DISCUSSION TITLE
How to create a specific cycle interval for partitions?

### DISCUSSION THREAD

**Discussion by JulesVerne22** [dagster-io/dagster#19647] [Q&A]
A lot of flight data uses a 28 day cycle, so I'm trying to partition the data in that way as well. However, I can't seem to figure out how to create a 28 day partition. I know I can create a partition that goes on the 28th day of the month using a cron schedule, but that does not fit my needs. I also have a way of determining the 28 day start dates, but I don't know how to have the partitions update when the next 28th day comes. Any help would be greatly appreciated!

**Comment by clayheaton** [comment]
You will want to use [`TimeWindowPartitionsDefinition`](https://docs.dagster.io/_apidocs/partitions#dagster.TimeWindowPartitionsDefinition) and the `cron_schedule` parameter to specify the period. 

For example, Here's how I defined a quarterly partition starting on April 1 2023 (our fiscal year starts then):

```python
@asset(
    required_resource_keys={REQUIRED_RESOURCE},
    partitions_def=TimeWindowPartitionsDefinition(
        cron_schedule="0 0 1 */3 *", start="2023-04-01", fmt="%Y-%m-%d"
    ),
    metadata={
        "partition_expr": "START_DATE",
    }
)
def TEST_QUARTERLY_PARTITIONS(context: AssetExecutionContext) -> pd.DataFrame:
    """
    Simple test and POC of TimeWindowPartitionsDefinition.
    """
    time_window = context.partition_time_window
    start_date = time_window.start.to_date_string()
    end_date = time_window.end.to_date_string()

    # Now do whatever you need to do

    return df
```

**Reply by JulesVerne22** [reply]
There is no way to define every 28 days via a cron schedule as far as I know. If you try to do something like */28 you'll get the 28th day of every month which is not what I'm looking for.

**Reply by clayheaton** [reply]
Maybe you'll need to skip Dagster's partitioning system then?

One idea would have an asset that looks up the last run date of the job. If it detects that it has been 28 days since the last running, then it fires off a run request to have the job run again. In your Definitions, set the "meta job" to run every day - it should only trigger the 28-day job when the appropriate criteria are met? Then the last step in the 28-day job would be to update the table/metadata with information about when it last ran. If you're trying to append to a SQL table, you'll have to write the SQL to append to an existing table instead of returning a dataframe that overwrites the table.

**Reply by mlarose** [reply]
Hi @JulesVerne22, 

For the scheduling aspect of it, you could set a job that is executed daily and use the should_execute callback property of [Schedule](https://docs.dagster.io/_apidocs/schedules-sensors#schedules) to add your business logic to evaluate whether or not this is the end of a 28 day cycle, and return False when it isn't.
